server/src/main/java/com/linecorp/centraldogma/server/internal/replication/ZooKeeperCommandExecutor.java

```diff
@@ -714,37 +757,92 @@ public void childEvent(CuratorFramework unused, PathChildrenCacheEvent event) th
         }
 
-        return () -> {
-            try {
-                mtx.release();
-            } catch (Exception ignored) {
-                // Ignore.
-            }
-        };
+        if (writeLock != null) {
+            scheduleWriteLockRelease(writeLock, mtx, executionPath);
+        }
+
+        return () -> safeRelease(mtx);
     }
 
-    @VisibleForTesting
-    static String path(String... pathElements) {
+    private static void safeRelease(InterProcessMutex mtx) {
+        try {
+            mtx.release();
+        } catch (Exception ignored) {
+            // Ignore.
+        }
+    }
+
+    @Nullable
+    private WriteLock acquireWriteLock(PushCommand command) throws Exception {
+        if (command.projectName().equals(INTERNAL_PROJ) ||
+            command.repositoryName().equals(Project.REPO_DOGMA)) {
+            // Do not check quota for internal project and repository.
+            return null;
+        }
+
+        final RepositoryMetadata meta;
+        try {
+            meta = metadataService.getRepo(command.projectName(), command.repositoryName()).get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new CentralDogmaException("Unexpected exception caught while retrieving " +
+                                            RepositoryMetadata.class.getSimpleName(), e);
+        }
+
+        QuotaConfig writeQuota = meta.writeQuota();
+        if (writeQuota == null) {
+            writeQuota = this.writeQuota;
+        }
+        if (writeQuota == null) {
+            return null;
+        }
+
+        final int requestQuota = writeQuota.requestQuota();
+        final Entry<InterProcessSemaphoreV2, SettableSharedCount> semaphoreAndCount =
+                semaphoreMap.computeIfAbsent(command.executionPath(), k -> {
+                    final SettableSharedCount cnt = new SettableSharedCount(requestQuota);
+                    return new SimpleImmutableEntry<>(
+                            new InterProcessSemaphoreV2(curator, absolutePath(QUOTA_PATH, k), cnt), cnt);
+                });
+
+        final SettableSharedCount count = semaphoreAndCount.getValue();
+        if (count.getCount() != requestQuota) {
+            count.setCount(requestQuota);
+        }
+        final InterProcessSemaphoreV2 semaphore = semaphoreAndCount.getKey();
+        final Lease lease = semaphore.acquire(200, TimeUnit.MILLISECONDS);
+        return new WriteLock(semaphore, lease, writeQuota);
+    }
+
+    private void scheduleWriteLockRelease(WriteLock writeLock, InterProcessMutex mtx, String executionPath) {
+        final Lease lease = writeLock.lease;
+        final QuotaConfig writeQuota = writeLock.writeQuota;
+        if (lease == null) {
+            safeRelease(mtx);
+            throw new TooManyRequestsException("commits", executionPath, writeQuota.requestQuota());
+        } else {
+            quotaExecutor.schedule(() -> writeLock.semaphore.returnLease(lease),
```

Comment:
What happens if this schedule is canceled when stopping this commandExectuor?
Shouldn't we also have the `writeLock` in the map and release it when this commandExectuor is stopped?

Reply:
I read the Javadoc of `InterProcessSemaphoreV2` that says if a client is closed, all acquired leases are closed too.
https://github.com/apache/curator/blob/95f147f8a864375744a566af3d550659f512728b/curator-recipes/src/main/java/org/apache/curator/framework/recipes/locks/InterProcessSemaphoreV2.java#L72-L73

Reply:
Then we don't need to close. Thanks! 😄